djstripe/models.py

```diff
@@ -357,35 +368,92 @@ def can_charge(self):
+        
+    def matching_stripe_subscription(self, subscription, cu=None):
+        """
+        Look up the Stripe subscription matching this subscription, using the
+        Stripe ID. If multiple subscriptions are not allowed, the subscription
+        parameter is ignored, and we just return the first (and presumably only)
+        Stripe subscription.
+        """
+        cu = cu or self.stripe_customer
+        sub = None
+        if cu.subscriptions.count > 0:
+            if not Customer.allow_multiple_subscriptions:
+                sub = cu.subscriptions.data[0]
+            else:
+                if subscription:
+                    matching = [s for s in cu.subscriptions.data if s.id == subscription.stripe_id]
+                    if matching:
+                        sub = matching[0]
+        return sub
+
+    def cancel_subscription(self, at_period_end=True, subscription=None):
+        if Customer.allow_multiple_subscriptions and not subscription:
+            raise SubscriptionApiError(
+                "Must specify a subscription to cancel"
+            )
+        if not subscription:
+            if self.subscriptions.count() == 0:
+                raise SubscriptionCancellationFailure(
+                    "Customer does not have current subscription"
+                )
+            subscription = self.subscriptions.all()[0]
 
         try:
             """
             If plan has trial days and customer cancels before trial period ends,
             then end subscription now, i.e. at_period_end=False
             """
-            if self.current_subscription.trial_end and self.current_subscription.trial_end > timezone.now():
+            if subscription.trial_end and subscription.trial_end > timezone.now():
                 at_period_end = False
-            sub = self.stripe_customer.cancel_subscription(at_period_end=at_period_end)
+            cu = self.stripe_customer
+            sub = None
+            if cu.subscriptions.count > 0:
+                if Customer.allow_multiple_subscriptions:
+                    matching = [sub for sub in cu.subscriptions.data if sub.id == subscription.stripe_id]
+                    if matching:
+                        sub = matching[0]
+                else:
+                    sub = cu.subscriptions.data[0]
+            if sub:
+                sub = sub.delete(at_period_end=at_period_end)
+                subscription.status = sub.status
+                subscription.cancel_at_period_end = sub.cancel_at_period_end
+                subscription.current_period_end = convert_tstamp(sub, "current_period_end")
+                subscription.canceled_at = convert_tstamp(sub, "canceled_at") or timezone.now()
+                if not at_period_end:
+                    subscription.ended_at = subscription.canceled_at
+                subscription.save()
+                cancelled.send(sender=self, stripe_response=sub)
+            else:
+                """
```

Comment:
I'm assuming this was added as a fix for #139. Nice - seems like a better place to put it than where I put my solution. I suggest using Subscription.STATUS_CANCELLED instead of hard coding "canceled" though.